django/db/migrations/executor.py

```diff
@@ -224,12 +224,17 @@ def _migrate_all_backwards(self, plan, full_plan, fake):
         # remaining applied migrations.
         last_unapplied_migration = plan[-1][0]
         state = states[last_unapplied_migration]
+        # Avoid mutating state with apps rendered as it's an expensive
+        # operation.
+        del state.apps
         for index, (migration, _) in enumerate(full_plan):
             if migration == last_unapplied_migration:
                 for migration, _ in full_plan[index:]:
                     if migration in applied_migrations:
                         migration.mutate_state(state, preserve=False)
                 break
+        # Re-render apps.
+        state.apps
```

Comment:
FWIW we could avoid forcing this to happen and just let the next access render it on demand. It might not always be necessary.

Reply:
Good point :dart:

Reply:
While the `migrate` command [will systematically need it in order to emit post-migrate signals as of today](https://github.com/django/django/blob/b85ab169dd1335afa7a1cbe824f53e5825044009/django/core/management/commands/migrate.py#L363-L390) it might not be the case in the future.

For example, we might avoid the whole rendering if there are no listeners for the `post_migrate` signal assuming we find a way to automatically create contenttypes + permissions another way.